File: StudyPlans/DataStructure_I/53.MaximumSubarray.ts
// Brute Force | Trying out every possible sub-array in nums
// Time exceeded 
// function maxSubArray(nums: number[]): number {
//   let maxSum = -Infinity

//   for (let i = 0; i < nums.length; i++) {
//     let currentSum = 0
//       for (let j = i; j < nums.length; j++) {
//         currentSum += nums[j] 
//         maxSum = Math.max(maxSum, currentSum)
//       }
//   }

//   return maxSum
// };

// console.log(maxSubArray([1,2.-7,5,-3,6,1,3,-100]));

// Dynamic Programing | Tabulation
function maxSubArray(nums: number[]): number {
  let local = 0
  let global = -Infinity

  for (const num of nums) {
    local = Math.max(num, local + num)
    if (local > global) { global = local }
  }

  return global
}

console.log(maxSubArray([1,2.-7,5,-3,6,1,3,-100]));

File: StudyPlans/DataStructure_I/53.MaximumSubarray..py
import math

class Solution:
  def maxSubArray(self, nums) -> int:
    maxSum = -math.inf

    for i in range(len(nums)):
      currentSum = 0
      for j in range(i, len(nums)):
        currentSum += nums[j]
        maxSum = max(maxSum, currentSum)
  
    return maxSum

# Dynamic Programing | Memoization
class SolutionII:
    def maxSubArray(self, nums):
        dp = [[0]*len(nums) for i in range(2)]
        dp[0][0], dp[1][0] = nums[0], nums[0]
        for i in range(1, len(nums)):
            dp[1][i] = max(nums[i], nums[i] + dp[1][i-1])
            dp[0][i] = max(dp[0][i-1], dp[1][i])
        return dp[0][-1]
